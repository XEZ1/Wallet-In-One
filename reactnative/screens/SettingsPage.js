import React, { useState, useEffect } from 'react';
import {
  View,
  Text,
  Switch,
  StyleSheet,
  TouchableOpacity,
  ScrollView,
  Appearance,
  AsyncStorage,
  FlatList,
  TouchableHighlight
} from 'react-native';

import * as SecureStore from 'expo-secure-store';
import { logout } from '../authentication';
import { useContext } from 'react';
import { userContext } from '../data';
import { useTheme } from 'reactnative/src/theme/ThemeProvider'
import { sendNotification,sendThemeNotification } from "./SendNotification";
import {styles} from 'reactnative/screens/All_Styles.style.js'

import Icon from 'react-native-vector-icons/Feather';

export default function SettingsPage ({ navigation }) {
  const [notifications, setNotifications] = useState(false);

  const [user, setUser] = useContext(userContext)

  const {dark, colors, setScheme, update} = useTheme();
  
  //const [isDarkModeEnabled, setIsDarkModeEnabled] = useState(false);

  const stylesInternal = StyleSheet.create({
    container: {
      flex: 1,
    },
    title: {
      fontSize: 24,
      fontWeight: 'bold',
      marginBottom: 20
    },
    switchContainer: {
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'space-between',
      width: '80%'
    },
    logoutButton: {
      marginTop: 20
    },
    aboutUs: {
      backgroundColor: 'red',
      color: 'black',
      width: "30%",
      borderRadius: 25,
      textAlign: 'center',
      fontWeight: 'bold',
      padding: "2%",
      fontSize:  17,
      marginTop: '10%',
      alignSelf: 'center'
    },
    developers: {
      backgroundColor: 'black',
      color: 'red',
      width: "40%",
      borderRadius: 25,
      textAlign: 'center',
      fontWeight: 'bold',
      padding: "2%",
      fontSize:  17,
      marginTop: '10%',
      alignSelf: 'center',
    },
    button: {
      width: "75%",
      borderRadius: 25,
      textAlign: 'center',
      fontWeight: 'bold',
      marginTop: '4%',
      paddingHorizontal: "12%",
      paddingVertical: "2%",
      fontSize:  20,
    },
    list: {
      backgroundColor: colors.background,
      width: "80%",
      borderWidth: 1,
      borderRadius: 4,
      borderColor: 'lightgrey',
    },
    item: {
      padding: 10,
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'space-between',
      borderBottomColor: '#ddd',
      borderBottomWidth: 1,
      height: 50
    },
    last_item: {
      borderBottomWidth: 0,
    },
  });

  const [selectedValue, setSelectedValue] = useState(null);

  function Item({label, value, style = stylesInternal.item}) {
    return <TouchableHighlight onPress={() => changeTheme(value)} underlayColor={dark?'#666':'#ddd'}>
      <View style={style}>
        <Text style={{color: colors.text}}>{label}</Text>
        {selectedValue === value && (<Icon name={'check'} size={24} color={'green'} />)}
      </View>
    </TouchableHighlight>;
  }

  //Load notification setting
  useEffect(() => {
      async function loadNotificationSetting() {

        const initialValue = await SecureStore.getItemAsync('darkModeSettings');
        setSelectedValue(initialValue);

        try {
          const savedNotificationSetting = await SecureStore.getItemAsync('notificationSetting');
        
          setNotifications(savedNotificationSetting === 'true');
        } catch(e) {
          console.warn("Couldn't load notification setting")
        }
      }
      loadNotificationSetting();
  }, []);

  // useEffect(() => {
  //   // Load user's saved dark mode setting on component mount
  //   async function loadDarkModeSettings() {
  //     const savedDarkModeSettings = await AsyncStorage.getItem('darkModeSettings');
  //     setIsDarkModeEnabled(savedDarkModeSettings === 'true');
  //     if (savedDarkModeSettings === 'true') {
  //         setScheme('dark');
  //     } else {
  //         setScheme('light');
  //     }
  //   }
  //   loadDarkModeSettings();
  // }, []);


  //Toggle and save notification setting
  const toggleNotifications = async () => {
      setNotifications(previousState => !previousState);
      const notificationSettings = (!notifications).toString();
      await  SecureStore.setItemAsync('notificationSettings', (!notifications).toString());
      sendNotification(notificationSettings);

  };

  //Toggle and save theme setting
  const changeTheme = async (theme) => {
      setSelectedValue(theme)
      update(theme)
      if (theme){
        await SecureStore.setItemAsync('darkModeSettings', theme);
      }
      else{
        await SecureStore.deleteItemAsync('darkModeSettings');
      }
      const notificationSettings = notifications.toString();
      sendThemeNotification(notificationSettings);
  };

  
  return (
    
    <ScrollView
      contentContainerStyle={{
        flexGrow : 1,
        justifyContent: 'center',
        alignItems: 'center',
        paddingBottom: 20,
        backgroundColor: colors.background,
        }}
      style={styles(dark, colors).container}
    >
      <Text style={[styles(dark, colors).textBold, {fontSize: 24, marginBottom: 20}]}>Notifications</Text>
      <View style={stylesInternal.switchContainer}>
        <Text style={styles(dark, colors).text}>Receive notifications</Text>
        <Switch
          trackColor={{ false: "#767577", true: "#81b0ff" }}
          thumbColor={notifications ? colors.primary : "#f4f3f4"}
          onValueChange={toggleNotifications}
          value={notifications}
        />
      </View>

      <Text style={[styles(dark, colors).textBold, {fontSize: 24, marginBottom: 20}]}>Themes</Text>

      {/* <View style={stylesInternal.switchContainer}>
        <Text style={styles(dark, colors).text}>Dark Mode (Beta)</Text>
        <Switch
          trackColor={{ false: "#767577", true: "#81b0ff" }}
          thumbColor={dark ? colors.primary : "#f4f3f4"}
          onValueChange={toggleTheme}
          value={dark}
        />
      </View> */}

      <View style={stylesInternal.list}>
        <Item label={'System Default'} value={null} />
        <Item label={'Light Mode'} value={'false'}/>
        <Item label={'Dark Mode'} value={'true'} style={[stylesInternal.item, stylesInternal.last_item]}/>
      </View>

      <TouchableOpacity
        onPress={()=>{logout(user, setUser)}}
      >
        <Text style={[{backgroundColor: colors.primary}, {color: "white"}, stylesInternal.button]}>Logout</Text>
      </TouchableOpacity>

      {/* <TouchableOpacity
        onPress={() => navigation.navigate('About Us')}
      >
        <Text style={stylesInternal.aboutUs}>About Us</Text>
      </TouchableOpacity>

      <TouchableOpacity
        onPress={() => navigation.navigate('Developer Info')}
      >
        <Text style={stylesInternal.developers}>Meet the team!</Text>
      </TouchableOpacity> */}
    </ScrollView>
  );
}