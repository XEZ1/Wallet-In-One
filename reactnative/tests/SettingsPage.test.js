import React from 'react';
import SettingsPage from 'reactnative/screens/SettingsPage';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react-native';

  jest.mock('react', () => {
      const ActualReact = jest.requireActual('react')
      return {
        ...ActualReact,
        useContext: () => (ActualReact.useState({'signedIn': false})),
      }
    })

  it('snapshot test for settings page', () => {
      const navigate = jest.fn();
      const settingsPage = renderer.create(<SettingsPage navigation={{ navigate }} />);
      expect(settingsPage).toMatchSnapshot();         
    });

  it('settings page pressing log out button test', () => {
      const navigate = jest.fn();
      const settingsPage = render(<SettingsPage navigation={ {navigate} }/>);
      fireEvent.press(screen.getByText('Logout'));
      expect(settingsPage).toMatchSnapshot();
  });

  it('setting page dark mode toggle', () => {
    const navigate = jest.fn();
    const settingsPage = render(<SettingsPage navigation={ {navigate} }/>);
    fireEvent.press(screen.getByText('Dark Mode'));
    expect(settingsPage).toMatchSnapshot();
  });