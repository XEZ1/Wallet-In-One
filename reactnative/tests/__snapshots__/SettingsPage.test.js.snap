// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test for settings page 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "alignItems": "center",
      "flexGrow": 1,
      "justifyContent": "center",
      "paddingBottom": 20,
    }
  }
  style={
    {
      "backgroundColor": undefined,
      "flex": 1,
    }
  }
>
  <View>
    <Text
      style={
        [
          {
            "color": undefined,
            "fontWeight": "700",
          },
          {
            "fontSize": 24,
            "marginBottom": 20,
          },
        ]
      }
    >
      Notifications
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "width": "80%",
        }
      }
    >
      <Text
        style={
          {
            "color": undefined,
          }
        }
      >
        Receive notifications
      </Text>
      <RCTSwitch
        accessibilityRole="switch"
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onTintColor="#81b0ff"
        style={
          {
            "height": 31,
            "width": 51,
          }
        }
        thumbTintColor="purple"
        tintColor="#767577"
        value={true}
      />
    </View>
    <Text
      style={
        [
          {
            "color": undefined,
            "fontWeight": "700",
          },
          {
            "fontSize": 24,
            "marginBottom": 20,
          },
        ]
      }
    >
      Themes
    </Text>
    <View
      style={
        [
          {
            "backgroundColor": undefined,
            "borderColor": "lightgrey",
            "borderRadius": 4,
            "borderWidth": 1,
            "width": "80%",
          },
        ]
      }
    >
      <View
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#ddd",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "space-between",
              "padding": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": undefined,
              }
            }
          >
            System Default
          </Text>
          <Text />
        </View>
      </View>
      <View
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderBottomColor": "#ddd",
              "borderBottomWidth": 1,
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "space-between",
              "padding": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": undefined,
              }
            }
          >
            Light Mode
          </Text>
        </View>
      </View>
      <View
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "borderBottomColor": "#ddd",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "height": 50,
                "justifyContent": "space-between",
                "padding": 10,
              },
              {
                "borderBottomWidth": 0,
              },
            ]
          }
        >
          <Text
            style={
              {
                "color": undefined,
              }
            }
          >
            Dark Mode
          </Text>
        </View>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "backgroundColor": undefined,
            },
            {
              "color": "white",
            },
            {
              "borderRadius": 25,
              "fontSize": 20,
              "fontWeight": "bold",
              "marginTop": "4%",
              "paddingHorizontal": "12%",
              "paddingVertical": "2%",
              "textAlign": "center",
              "width": "75%",
            },
          ]
        }
      >
        Logout
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
